### PR TITLE
OCPBUGS-44641: GCP placing *.apps record in wrong managed zone

### PR DESCRIPTION
**Pull request by barbacbd**
** The *.apps record is not being placed in the correct DNS Managed Zone. The existing managed zone should be used only during xpn installs. The ingress operator was supplied with the incorrect managed zone information.

### CODE REVIEW

File: pkg/asset/manifests/dns.go [hunk 1]
```diff
@@ -67,6 +67,8 @@ func (*DNS) Dependencies() []asset.Asset {
 }
 
 // Generate generates the DNS config and its CRD.
+//
+//nolint:gocyclo
```
Comment (r4f4) [inline pkg/asset/manifests/dns.go:71]: Nit: it's nice to add a reason why it's ok to silence the linter.

**Review by r4f4 — COMMENTED** [review]
/lgtm

File: pkg/asset/manifests/dns.go (hunk 1 above)
Comment (patrickdillon) [reply pkg/asset/manifests/dns.go:71]: The way to fix this linting issue is to move the code into its own function, especially if you make the changes to check the `PrivateVisabilityConfig` of the zone

**Review by patrickdillon — COMMENTED** [review]
I agree with this change: we should only do this for xpn installs, but it doesn't fix the bug in question. To fix the bug, you need to look at the `zone.PrivateVisibilityConfig.Networks` on the returned zone and ensure that it contains the network(s?) from the install config.

File: pkg/asset/manifests/dns.go [hunk 2]
```diff
@@ -229,6 +225,36 @@ func (d *DNS) Generate(ctx context.Context, dependencies asset.Parents) error {
...
+// occurs, a precreated zone may be used. If a zone is found (in this instance), then the zone should be paired with
+// the network that is supplied through the install config (when applicable).
+func GetGCPPrivateZoneName(ctx context.Context, client *icgcp.Client, installConfig *installconfig.InstallConfig, clusterID string) (string, error) {
+	privateZoneID := fmt.Sprintf("%s-private-zone", clusterID)
+	if installConfig.Config.GCP.NetworkProjectID != "" {
+		zone, err := client.GetDNSZone(ctx, installConfig.Config.GCP.ProjectID, installConfig.Config.ClusterDomain(), false)
+		if err != nil {
+			return "", fmt.Errorf("failed to get private zone for %q: %w", installConfig.Config.BaseDomain, err)
+		}
+		if zone != nil {
```
Comment (r4f4) [inline pkg/asset/manifests/dns.go:238]: So the `GetDNSZone` call can return `nil` even if there weren't any errors?

Comment (barbacbd) [reply pkg/asset/manifests/dns.go:238]: Yes, it returns nil when there are no zones found. It is not an error to find no zones. This is an indicator that a zone must be created.

File: pkg/asset/manifests/dns.go [hunk 3]
```diff
@@ -229,6 +225,36 @@ func (d *DNS) Generate(ctx context.Context, dependencies asset.Parents) error {
...
+			if installConfig.Config.GCP.Network != "" {
+				expectedNetworkURL := fmt.Sprintf(
+					"https://www.googleapis.com/compute/v1/projects/%s/global/networks/%s",
+					installConfig.Config.GCP.NetworkProjectID,
+					installConfig.Config.GCP.Network,
+				)
```
Comment (r4f4) [inline pkg/asset/manifests/dns.go:244]: This is repeated in the capi code so maybe it's a candidate for a small shared function?

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Pull request merged**